### PR TITLE
Tag AppSec on Rack top-level span

### PR DESCRIPTION
**Pull request by lloeki**
**What does this PR do?**
Tag AppSec on Rack top-level span

**Motivation**
AppSec discovers service name based on top-level span. Tagging at trace level results in tags on the trace root span, which is usually Rack's top-level span.

When request queuing is enabeld, though, a synthetic span is added in front, representing a service outside the application. This tag then gets mistakenly tagged with AppSec information, resulting in much confusion.

The situation gets even trickier with partial flushing, as partially flushed spans may receive a trace tag.

Ensuring we tag the proper span should solve both issues.

**Additional Notes**

Starting this PR as draft since I am still figuring out how to handle `Datadog::Kit` tagging in a backward compatible manner.

**How to test the change?**

Add `request_queuing: true` to Rack or Rails tracing integration, then:

```
curl -vvv -H 'User-Agent: Arachni/v2' -H "X-Request-Start: $(ruby -e 'puts "%d" % (Time.now.to_f * 1000)')" -H "X-Queue-Start: $(ruby -e 'puts "%d" % (Time.now.to_f * 1000)')" 'http://user:pass@127.0.0.1:9292/'
```

### CODE REVIEW

File: lib/datadog/kit/appsec/events.rb [hunk 1]
```diff
@@ -14,56 +14,107 @@ module Events
...
+            trace = appsec_scope.trace
+            span = appsec_scope.span
+          end
+
+          trace ||= Datadog::Tracing.active_trace
+          span ||= trace.active_span || Datadog::Tracing.active_span
+
+          if trace.trace_id != span.trace_id
+            raise ArgumentError, "span #{span.span_id} does not belong to trace #{trace.trace_id}"
+          end
```
Comment (GustavoCaso) [inline lib/datadog/kit/appsec/events.rb:38]: We should probably extra this to a separate helper method within appsec 😄

File: lib/datadog/kit/appsec/events.rb [hunk 2]
```diff
@@ -14,56 +14,101 @@ module Events
...
+        def self.track_login_success(trace = nil, span = nil, user:, **others)
+          if (appsec_scope = Datadog::AppSec.active_scope)
+            trace = appsec_scope.trace
+            span = appsec_scope.service_entry_span
+          end
+
+          trace ||= Datadog::Tracing.active_trace
+          span ||= trace.active_span || Datadog::Tracing.active_span
+
+          raise ArgumentError, "span #{span.span_id} does not belong to trace #{trace.id}" if trace.id != span.trace_id
```
Comment (GustavoCaso) [inline lib/datadog/kit/appsec/events.rb:36]: Should we raise or log an message letting the user that they have done something odd?

File: lib/datadog/appsec/contrib/rack/request_body_middleware.rb [hunk 3]
```diff
@@ -17,7 +17,7 @@ def initialize(app, opt = {})
           end
 
           def call(env)
-            context = env['datadog.waf.context']
+            context = env['datadog.appsec.scope']
```
Comment (lloeki) [inline lib/datadog/appsec/contrib/rack/request_body_middleware.rb]: Thanks @GustavoCaso !

File: lib/datadog/kit/appsec/events.rb (hunk 1 above)
Comment (lloeki) [reply lib/datadog/kit/appsec/events.rb]: Yeah, I went for the quick fix route here, as I find it awkward currently to hide that code behind the wrong abstraction.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb [hunk 4]
```diff
@@ -70,33 +70,37 @@ def call(env)
...
+              Datadog::AppSec::Scope.deactivate_scope
             end
           end
           # rubocop:enable Metrics/AbcSize,Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity,Metrics/MethodLength
 
           private
 
+          def active_scope(env)
+            env['datadog.appsec.scope']
+          end
```
[the remark this answers is not in the document]
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb]: Agreed. Let's address this later and get the release done.

File: lib/datadog/kit/appsec/events.rb (hunk 2 above)
Comment (lloeki) [reply lib/datadog/kit/appsec/events.rb:36]: As with the active scope (prev. active context) sanity checks, this should not happen, ever. If we swallow it we'll never be aware.

**Pull request merged**